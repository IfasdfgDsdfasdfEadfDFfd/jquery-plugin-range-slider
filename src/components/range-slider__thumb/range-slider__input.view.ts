import { View } from '@core';
import { memo } from 'core/utils';

@memo(['updateColor'])
class Thumb extends View<ThumbProps> {
  attrs = {
    class: 'range-slider__thumb',
  };
  children = {
    marker: new ThumbMarker(),
  };

  render({ markerText, color }: ThumbProps): void {
    this.updateColor(color);

    this.children.marker.render({ text: markerText, color });
  }

  updateColor(newColor: string): void {
    this.nativeElement.style.setProperty('border-color', newColor);
  }
}

@memo(['updateText', 'updateColor'])
class ThumbMarker extends View<ThumbMarkerProps> {
  attrs = {
    class: 'range-slider__thumb__marker',
  };

  render({ text, color }: ThumbMarkerProps): void {
    this.updateText(text);
    this.updateColor(color);
  }

  updateText(newText: string): void {
    this.nativeElement.textContent = newText;
  }

  updateColor(newColor: string): void {
    this.nativeElement.style.setProperty('color', newColor);
  }
}

export { Thumb, ThumbMarker };
