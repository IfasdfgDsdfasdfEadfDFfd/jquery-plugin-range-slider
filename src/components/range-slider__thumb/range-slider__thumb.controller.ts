import { Controller } from '@core';

class ThumbController extends Controller {
  mapState({ root, input }: RangeSliderModelData): ThumbProps {
    const { min, max, values } = input;

    return {
      markerText: values[0].toString(),
      color: root.color,
      positionOffset: this.calcPositionOffset({ min, max, value: values[0] }),
    };
  }

  calcPositionOffset({ min, max, value }: { min: number; max: number; value: number }): number {
    return ((value - min) / (max - min)) * 100;
  }
}

export { ThumbController };
