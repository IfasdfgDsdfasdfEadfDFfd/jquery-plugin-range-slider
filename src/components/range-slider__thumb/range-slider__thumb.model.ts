import { Model } from '@core';

class ThumbModel extends Model<ThumbModelData> {
  name = 'thumb';
  data = {};
}

export { ThumbModel };
