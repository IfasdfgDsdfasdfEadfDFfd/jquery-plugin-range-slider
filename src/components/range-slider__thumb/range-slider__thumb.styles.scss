@import '../../variables.scss';

.range-slider__thumb {
  position: absolute;
  top: $thumb-height * .1;

  background-color: $primary-color;

  width: $thumb-width * .8;
  height: $thumb-height * .8;

  border: $thumb-border-width solid #fff;
  border-radius: $thumb-border-radius;

  transition: all .2s ease;

  opacity: 1;
  z-index: 1;
}

.range-slider__thumb--hovered,
.range-slider__thumb--focused {
  top: 0;
  width: $thumb-width;
  height: $thumb-height;

  transition: all .2s ease;
}

.range-slider__thumb--hidden {
  opacity: 0;
}

.range-slider__thumb__marker {
  position: relative;

  display: flex;
  justify-content: center;
  align-items: center;

  color: #fff;

  margin-top: -3rem;
  padding: .3rem 0;

  background-color: $primary-color;
}
