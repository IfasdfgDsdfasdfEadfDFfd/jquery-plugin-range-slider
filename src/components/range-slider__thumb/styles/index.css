.range-slider__thumb {
  position: absolute;
  width: var(--thumb-width);
  height: var(--thumb-height);
  background-color: var(--primary-color);
}
