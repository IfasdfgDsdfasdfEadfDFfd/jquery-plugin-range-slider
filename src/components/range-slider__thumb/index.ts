import './range-slider__thumb.styles.scss';

export { Thumb } from './range-slider__thumb.view';
