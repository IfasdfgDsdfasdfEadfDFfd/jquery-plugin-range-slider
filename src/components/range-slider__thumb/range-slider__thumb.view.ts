import { View } from '../../core';

import styles from '../../exports.scss';

class Thumb extends View {
  hidingClassName = 'range-slider__thumb--hidden';
  focusClassName = 'range-slider__thumb--focused';
  hoverClassName = 'range-slider__thumb--hovered';

  marker!: ThumbMarker;
  lastValues!: {
    max: number;
    min: number;
    value: number;
    prefix: string;
    postfix: string;
    displayValue: number | string;
  };
  lastColor!: string;
  isFocused = false;

  constructor() {
    super({
      tag: 'div',
      attrs: { class: 'range-slider__thumb' },
      children: [new ThumbMarker()],
    });
    this.marker = this.children[0] as ThumbMarker;
  }

  positioning(
    max: number,
    min: number,
    value: number,
    prefix = '',
    postfix = '',
    displayValue?: number | string,
  ): void {
    displayValue = displayValue || value;
    this.lastValues = { max, min, value, prefix, postfix, displayValue };

    const thumbWidth =
      <number>this.element.clientWidth + parseInt(styles.thumbWidth);
    const sliderWidth = <number>this.element.parentElement?.clientWidth;

    const offset = this.getOffset({
      selfWidth: thumbWidth,
      parentWidth: sliderWidth,
      value,
      max,
      min,
    });

    this.element.style.setProperty('left', `${offset}%`);

    this.marker.value = `${prefix}${displayValue}${postfix}`;
  }

  set primaryColor(value: string) {
    if (value !== this.lastColor) {
      this.element.style.setProperty('background-color', value);
      this.marker.element.style.setProperty('background-color', value);
      this.lastColor = value;
    }
  }

  positionCorrection(): void {
    const { max, min, value, prefix, postfix, displayValue } = this.lastValues;
    this.positioning(max, min, value, prefix, postfix, displayValue);
  }

  onFocus(): void {
    this.colorReset();
  }

  onHover(): void {
    this.colorReset();
  }

  colorReset(): void {
    if (this.isFocused || this.isHovered) {
      this.element.style.removeProperty('background-color');
      this.marker.element.style.removeProperty('background-color');
      this.element.style.setProperty('border-color', this.lastColor);
      this.marker.element.style.setProperty('color', this.lastColor);
      this.marker.element.style.setProperty('border-color', this.lastColor);
    } else {
      this.element.style.removeProperty('border-color');
      this.marker.element.style.removeProperty('border-color');
      this.marker.element.style.removeProperty('color');
      this.element.style.setProperty('background-color', this.lastColor);
      this.marker.element.style.setProperty('background-color', this.lastColor);
    }
  }

  getOffset({
    selfWidth,
    parentWidth,
    value,
    max,
    min,
  }: {
    [key: string]: number;
  }): number {
    const ratio = (value - min) / (max - min);
    const offsetPercent = 100 * ratio;
    const selfPercent = (selfWidth / parentWidth) * 100 * ratio;

    return offsetPercent - selfPercent;
  }
}

class ThumbMarker extends View {
  hidingClassName = 'range-slider__thumb__marker--hidden';
  _value!: string;

  constructor() {
    super({
      tag: 'div',
      attrs: { class: 'range-slider__thumb__marker' },
      children: [],
    });
  }

  set value(value: string) {
    this.replaceChildren([value]);

    const width = Math.max(
      value.length * parseInt(styles.rootFontSize),
      parseInt(styles.minThumbMarkerWidth) * parseInt(styles.rootFontSize),
    );
    this.element.style.setProperty('width', `${width}px`);

    const offset = -(
      (width - <number>this.element.parentElement?.clientWidth) /
      2
    );
    this.element.style.setProperty('margin-left', `${offset}px`);

    this._value = value;
  }

  positionCorrection() {
    this.value = this._value;
  }

  setVerticalMargin() {
    const offset = -this.element.clientWidth;
    this.element.style.setProperty('margin-top', `${offset}px`);
  }
}

export { Thumb };
