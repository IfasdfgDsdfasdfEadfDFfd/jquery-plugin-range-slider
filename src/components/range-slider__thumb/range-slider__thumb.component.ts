import { Component } from '@core';

import { Thumb } from './range-slider__thumb.view';
import { ThumbModel } from './range-slider__thumb.model';
import { ThumbController } from './range-slider__thumb.controller';

class ThumbComponent extends Component {
  view = new Thumb();
  model = new ThumbModel();
  controller = new ThumbController();
}

export { ThumbComponent };
