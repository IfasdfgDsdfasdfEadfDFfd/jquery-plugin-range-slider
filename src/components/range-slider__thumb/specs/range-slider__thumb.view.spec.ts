import { Thumb, ThumbMarker } from '../range-slider__thumb.view';

describe('Thumb', () => {
  let thumb: Thumb;
  const props: ThumbProps = {
    color: 'rgb(255, 255, 255)',
    markerText: 'marker text',
  };

  beforeEach(() => {
    thumb = new Thumb();
    thumb.init();
    thumb.render(props);
  });

  test('updateColor()', () => {
    const newColor = 'rgb(133, 133, 133)';
    expect(newColor).not.toEqual(props.color);

    thumb.updateColor(newColor);
    expect(thumb.nativeElement.style.getPropertyValue('border-color')).toEqual(newColor);
  });

  test('to match snapshot', () => {
    expect(thumb.nativeElement).toMatchSnapshot();
  });
});

describe('ThumbMarker', () => {
  let marker: ThumbMarker;

  const props: ThumbMarkerProps = {
    color: 'rgb(255, 255, 255)',
    text: 'marker text',
  };

  beforeEach(() => {
    marker = new ThumbMarker();
    marker.init();
    marker.render(props);
  });

  test('updateText()', () => {
    const newText = 'new marker text';

    expect(newText).not.toEqual(props.text);
    marker.updateText(newText);
    expect(marker.nativeElement.textContent).toEqual(newText);
  });

  test('updateColor()', () => {
    const newColor = 'rgb(133, 133, 133)';

    expect(newColor).not.toEqual(props.color);
    marker.updateColor(newColor);
    expect(marker.nativeElement.style.getPropertyValue('color')).toEqual(newColor);
  });

  test('to match snapshot', () => {
    expect(marker.nativeElement).toMatchSnapshot();
  });
});
