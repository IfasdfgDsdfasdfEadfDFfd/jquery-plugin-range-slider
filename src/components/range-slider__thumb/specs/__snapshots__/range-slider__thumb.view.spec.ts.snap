// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumb to match snapshot 1`] = `
<div
  class="range-slider__thumb"
  style="border-color: rgb(255, 255, 255);"
>
  <div
    class="range-slider__thumb__marker"
    style="color: rgb(255, 255, 255);"
  >
    marker text
  </div>
</div>
`;

exports[`ThumbMarker to match snapshot 1`] = `
<div
  class="range-slider__thumb__marker"
  style="color: rgb(255, 255, 255);"
>
  marker text
</div>
`;
